dll/win32/shell32/folders/CDesktopFolder.cpp

```diff
@@ -1102,6 +1032,48 @@ HRESULT WINAPI CDesktopFolder::CallBack(IShellFolder *psf, HWND hwndOwner, IData
     return SHELL32_DefaultContextMenuCallBack(psf, pdtobj, uMsg);
 }
 
+/*************************************************************************
+ * CDesktopFolderViewCB
+ */
+
+bool CDesktopFolderViewCB::IsProgmanHostedBrowser(IShellView *psv)
+{
+    FOLDERSETTINGS settings;
+    return SUCCEEDED(psv->GetCurrentInfo(&settings)) && (settings.fFlags & FWF_DESKTOP);
+}
+
+bool CDesktopFolderViewCB::IsProgmanHostedBrowser()
+{
+    enum { Uninitialized = 0, NotHosted, IsHosted };
+    C_ASSERT(Uninitialized == 0);
+    if (m_IsProgmanHosted == Uninitialized)
+        m_IsProgmanHosted = m_ShellView && IsProgmanHostedBrowser(m_ShellView) ? IsHosted : NotHosted;
+    return m_IsProgmanHosted == IsHosted;
+}
+
+HRESULT WINAPI CDesktopFolderViewCB::ShouldShow(IShellFolder *psf, PCIDLIST_ABSOLUTE pidlFolder, PCUITEMID_CHILD pidlItem)
```

Comment:
Why? Because Microsoft wanted XP to ship with the Recycle bin in the bottom right corner on the desktop with no other icons. Changing between the two start-menu types can change which icons are visible on the desktop. It's simply bad design.

The string "HideDesktopIcons" should give an idea of what this is about?